Use crypt_gensalt(), if available in libcrypt.

besser82 opened the pull request:
Most Linux distributions, including Fedora and RHEL 8, are shipping with libxcrypt >= 4.0.

Since that version of libxcrypt the provided family of crypt_gensalt() functions are able to use automatic entropy drawn from secure system ressources, like arc4random(), getentropy() or getrandom().

Anyways, the settings generated by crypt_gensalt() are always guaranteed to works with the crypt() function.

Using crypt_gesalt() is also needed to make proper use of newer hashing methods, like yescrypt, provided by libxcrypt.

***

See https://github.com/shadow-maint/shadow/pull/303#issuecomment-764847950 for reference.

besser82 commented:
@solardiz: This is PR going to make you happy.  =)

solardiz commented on libmisc/salt.c:
```diff
@@ -12,19 +12,314 @@
+	}
+
+	crypt_gensalt_rn(setting, rounds, NULL, 0,
+			 result, CRYPT_GENSALT_OUTPUT_SIZE);
```
Let's check the return value here. Return value is documented, what's in `result` upon error is not, and we're not in the same library with the implementation to make such assumptions.

besser82 replied:
Fixed in rebased commit.

solardiz reviewed (commented):
While I like where this is going (thank you!), I think it's just a first step, and I find it weird that with this PR's changes we at the same time require `crypt_gensalt_rn`, yet still have the old salt generation code somewhere else in shadow (dead code now?), the legacy per-hash-type settings, and the random rounds count in a range nonsense.

solardiz commented:
As you can see from the Owl and ALT patches I referenced in #303, longer-term there should be only two settings - `CRYPT_PREFIX` and `CRYPT_ROUNDS` - and nothing per-hash-type.

I understand this might not be right to "force" onto shadow users without warning (having their old settings stop working). However, in that case for consistency perhaps the dependency on libxcrypt 4.0+ should be optional for now?

Just thinking out loud.

solardiz commented:
Let's also hear from @ldv-alt on this.

besser82 commented:
@solardiz: The changes, I applied in this PR, are meant to be as transparent as possible to the end-user and/or distribution maintainer:

  * `crypt_gensalt()` is only used, if the system provides libxcrypt >= 4.0, otherwise the old code-path will be used.
  * There are no changes needed for existing configurations and such.

solardiz commented on libmisc/salt.c:
```diff
@@ -413,7 +645,7 @@ static /*@observer@*/const char *gensalt (size_t salt_size)
 	} else if (0 == strcmp (method, "YESCRYPT")) {
 		MAGNUM(result, 'y');
-		strcat(result, YESCRYPT_salt_cost((int *)arg));
+		strcat(result, YESCRYPT_salt_cost((long *)arg));
```
Looks like this was actually a bug being fixed here. So at least this bug fix should get in even if this PR maybe doesn't.

besser82 replied:
Indeed!

besser82 replied:
I've opened https://github.com/shadow-maint/shadow/pull/358 to keep this in mind, if the current PR doesn't get merged.

solardiz commented on libmisc/salt.c:
```diff
@@ -12,19 +12,321 @@
+		for (int i = 0; i < CRYPT_GENSALT_OUTPUT_SIZE; ++i) {
+			result[i] = '\0';
+		}
+	}
```
Why not call `crypt_gensalt_rn` from inside the `if`, so that we wouldn't need to introduce a variable? And why not use `memset` (if zeroizing the entire buffer is somehow needed, which I doubt)?

besser82 replied:
Fixed in rebased commit.

solardiz commented on libmisc/salt.c:
```diff
@@ -12,19 +12,321 @@
+	return result;
+}
+
+#else
```
Maybe add ` /*  USE_XCRYPT_GENSALT */` here, although I don't know if this project's style requires that we possibly negate the condition in such comment on `#else`.

besser82 replied:
Fixed in rebased commit.

solardiz commented:
BTW, why not use the non-thread-safe `crypt_gensalt` if we put the result in a static buffer anyway?

besser82 commented:
> BTW, why not use the non-thread-safe `crypt_gensalt` if we put the result in a static buffer anyway?

Wouldn't that just reorder zeroing the static buffer, need another variable for error checking, and require to strncpy the result of `crypt_gensalt` into the static buffer on success?

solardiz commented:
I was thinking that we'd simply return the pointer returned by `crypt_gensalt`, with no extra buffer nor copying. As to zeroing the buffer on failure, why? If we do that because we're concerned about any sensitive information being left in there (what information? just the hash type/settings?) then it's better done inside libxcrypt, not in the callers.

besser82 commented:
> I was thinking that we'd simply return the pointer returned by `crypt_gensalt`, with no extra buffer nor copying.

Okay, that makes sense.  Changed that in rebased commit.

ikerexxe reviewed (changes requested):
I've done a review of the c code and noted several items. Can you improve them?

By the way, I've seen that https://github.com/shadow-maint/shadow/pull/358 and this PR have several changes in common. Can you consider removing the common changes from this PR?

solardiz commented:
@ikerexxe I think most of your comments are about code/issues copied from elsewhere in the original file. My thinking was that improving this old code is out of scope of this PR, and increasing differences between the original vs. copied instances isn't obviously an improvement. Ideally, all of this would be obsoleted with more complete use of `crypt_gensalt` later, as I had suggested. At least the random rounds number in a range nonsense should be dropped soon, I think with a separate PR.

A possible improvement to this PR would be to avoid such code copying/duplication, if this can be done cleanly.

ldv-alt commented:
> Let's also hear from @ldv-alt on this.

ALT is using Owl's crypt_gensalt-based approach for almost 20 years now.
In ALT's shadow package [libmisc/salt.c](http://git.altlinux.org/gears/s/..git?p=shadow.git;a=blob;f=shadow/libmisc/salt.c) contains essentially the following (everything else is #if 0'ed):
```
const char *
crypt_make_salt(/*@null@*/const char *meth, /*@null@*/void *arg)
{
        char *retval;

        retval = crypt_gensalt(getdef_str("CRYPT_PREFIX"),
                        getdef_num("CRYPT_ROUNDS", 0), NULL, 0);
        if (!retval) {
                fprintf(stderr, "Unable to generate a salt, "
                                "check your CRYPT_PREFIX and CRYPT_ROUNDS settings.\n");
                exit(1);
        }

        return retval;
}
```

So from my point of view, the sooner all this legacy code is gone the better.

besser82 commented:
> I've done a review of the c code and noted several items. Can you improve them?

Working on it.


> By the way, I've seen that #358 and this PR have several changes in common. Can you consider removing the common changes from this PR?

This is intentional, as I've split out a bugfix into #358, that should not be overlooked in case this PR will *not* get merged.

solardiz commented:
I think if you keep the bugfix in this PR as well, then it should be a separate first commit (same as what's in the other PR).

besser82 commented on libmisc/salt.c:
```diff
@@ -12,19 +12,309 @@
+ * Other methods can be set with ENCRYPT_METHOD
+ *
+ * The method can be forced with the meth parameter.
+ * If NULL, the method will be defined according to the MD5_CRYPT_ENAB and
+ * ENCRYPT_METHOD login.defs variables.
```
Clarified in rebased commit.

besser82 commented on libmisc/salt.c:
```diff
@@ -12,19 +12,309 @@
+{
+	const char *method;
+	char setting[SETTING_SIZE];
+        unsigned long rounds = 0;
```
Fixed in rebased commit.

besser82 commented on libmisc/salt.c:
```diff
@@ -12,19 +12,309 @@
+#endif /* USE_YESCRYPT */
+
+/* Maximum size of the setting for crypt_gensalt(). */
+#define SETTING_SIZE 16
```
Fixed in rebased commit.

besser82 commented on libmisc/salt.c:
```diff
@@ -12,19 +12,309 @@
+		rounds = (unsigned long) *prefered_rounds;
+	}
+
+	/*
+	 * Sanity checks.
+	 * Use 19 as an upper bound for now,
+	 * because musl doesn't allow rounds >= 20.
+	 */
```
Removed, as obviously wrong, in rebased commit.

besser82 commented on libmisc/salt.c:
```diff
@@ -12,19 +12,309 @@
+		}
+
+		rounds = (unsigned long) shadow_random (min_rounds, max_rounds);
+	} else if (0 == *prefered_rounds) {
+		return 0;
```
Removed, as unneeded, in rebased commit.

besser82 commented on libmisc/salt.c:
```diff
@@ -12,19 +12,309 @@
+	}
+
+	if (rounds > B_ROUNDS_MAX) {
+		/* rounds = B_ROUNDS_MAX; */
```
This is still needed, as `B_ROUNDS_MAX` is defined to `31`, which is really the maximum rounds of the bcrypt hash method.

besser82 commented:
> I think if you keep the bugfix in this PR as well, then it should be a separate first commit (same as what's in the other PR).

Will fix it that way.

solardiz commented on libmisc/salt.c:
```diff
@@ -12,19 +12,302 @@
+/*
+ * Using the Prefix $2b$ to enable an anti-collision safety measure in musl libc.
+ * Negatively affects a subset of passwords containing the '\xff' character,
+ * which is not valid UTF-8 (so "unlikely to cause much annoyance").
```
When revised to say `$2b$`, this comment became factually wrong. The "anti-collision safety measure" was only enabled for `$2a$`.

solardiz replied:
To be clear, I think this comment should simply be dropped. In fact, it never made sense to deliberately enable the anti-collision safety measure for new password, like we do in the shadow package. That measure was only intended for authentication against old potentially-buggy hashes, not for new password hashes.

besser82 replied:
Fixed in rebased commit.

solardiz commented:
There's a typo in the function name in "Using crypt_gesalt() is also needed" in one of the commit messages.

besser82 commented:
> There's a typo in the function name in "Using crypt_gesalt() is also needed" in one of the commit messages.

Also fixed.

solardiz commented on libmisc/salt.c:
```diff
@@ -90,12 +90,11 @@ static void seedRNG (void)
- * which is not valid UTF-8 (so "unlikely to cause much annoyance").
+/*
+ * Using the Prefix $2b$, which is the preferred prefix to use for new bcrypt
+ * hahses for a long time now.
```
Typo in "hahses". I'd just drop the comment, really.

besser82 replied:
Fixed in rebased commit.

solardiz commented:
I find it weird that this PR adds as many as ~250 lines of code, most of those being copies of legacy code. There ought to be a clean way to rework it to keep and temporarily reuse the legacy code without having it duplicated. Perhaps the `#if`/`#else`/`#endif` could be moved further down, to be after the legacy functions that we temporarily reuse?

besser82 commented:
> I find it weird that this PR adds as many as ~250 lines of code, most of those being copies of legacy code. There ought to be a clean way to rework it to keep and temporarily reuse the legacy code without having it duplicated. Perhaps the `#if`/`#else`/`#endif` could be moved further down, to be after the legacy functions that we temporarily reuse?

Done so in rebased commit.

besser82 commented:
@solardiz I've given this a complete overhaul, and split out changes for code cleanup into a seperate commit before doing the implemetation of `crypt_gensalt()`.

Please have another look, if you can spare the time.

besser82 commented:
A breakdown of this PR can be found in the following PRs:

* Small bugfix for yescrypt impl.: https://github.com/shadow-maint/shadow/pull/358
* bcypt should use `$2b$`: https://github.com/shadow-maint/shadow/pull/360
* Cleanup of `libmisc/salt.c`: https://github.com/shadow-maint/shadow/pull/361

besser82 commented:
@ikerexxe Would you mind to have a second look on this?

solardiz commented on libmisc/salt.c:
```diff
@@ -12,33 +12,98 @@
 
+#if (defined CRYPT_GENSALT_IMPLEMENTS_AUTO_ENTROPY && \
+     CRYPT_GENSALT_IMPLEMENTS_AUTO_ENTROPY)
+/* Maximum size of the setting for crypt_gensalt(). */
```
This comment "Maximum size of the setting for crypt_gensalt()." looks irrelevant here.

besser82 replied:
Ahh, yes, that somehow slipped into there…  Will remove with next push.

besser82 replied:
Fixed.

solardiz reviewed (commented):
Great that the line count doesn't increase much anymore (except in the pot file), but now the "Sanitize code" commit makes somewhat invasive changes to legacy code, which I didn't fully review. Hopefully, they're OK.

ikerexxe commented:
> @ikerexxe Would you mind to have a second look on this?

Sure! I've been quite busy these days but I expect to have some time tomorrow.

ikerexxe reviewed (commented):
The code now is in a better shape and it's easier to understand. I've added several minor concerns but they should be easy to fix.

By the way, taking into account that you are already touching `salt.c` file, do you mind making the necessary changes to fix https://github.com/shadow-maint/shadow/issues/376?

besser82 commented on libmisc/salt.c:
```diff
@@ -156,196 +203,174 @@ static /*@observer@*/const char *SHA_salt_rounds (/*@null@*/int *prefered_rounds
+	}
+
+	/* Check if the result buffer is long enough. */
+	assert (GENSALT_SETTING_SIZE > buf_begin + 17);
```
Fixed. See #377 also.

besser82 commented on libmisc/salt.c:
```diff
@@ -156,196 +203,174 @@ static /*@observer@*/const char *SHA_salt_rounds (/*@null@*/int *prefered_rounds
+	/* Check if the result buffer is long enough. */
+	assert (GENSALT_SETTING_SIZE > buf_begin + 17);
+
+	(void) snprintf (buf + buf_begin, 18, "rounds=%lu$", rounds);
```
Fixed. See #377 also.

besser82 commented on libmisc/salt.c:
```diff
@@ -156,196 +203,174 @@ static /*@observer@*/const char *SHA_salt_rounds (/*@null@*/int *prefered_rounds
 
-	salt[0] = '\0';
+	/* Check if the result buffer is long enough. */
+	assert (GENSALT_SETTING_SIZE > buf_begin + 3);
```
Fixed. See #377 also.

besser82 commented on libmisc/salt.c:
```diff
@@ -156,196 +203,174 @@ static /*@observer@*/const char *SHA_salt_rounds (/*@null@*/int *prefered_rounds
-	salt[BCRYPT_SALT_SIZE] = '\0';
-
-	return salt;
+	(void) snprintf (buf + buf_begin, 4, "%2.2lu$", rounds);
```
Fixed. See #377 also.

besser82 commented:
> The code now is in a better shape and it's easier to understand. I've added several minor concerns but they should be easy to fix.

Done!


> By the way, taking into account that you are already touching `salt.c` file, do you mind making the necessary changes to fix #376?

See #377 also.

besser82 commented:
@hallyn Last one and we're done, I promise!  :P

ikerexxe commented:
LGTM. Thank you for your patches!

besser82 commented:
> LGTM. Thank you for your patches!

Please wait with merging!  I forgot to include <crypt.h>…

besser82 commented:
> > LGTM. Thank you for your patches!
> 
> Please wait with merging! I forgot to include <crypt.h>…

@hallyn Things should be fine for merging now.

besser82 commented:
@solardiz @ikerexxe Could you please set the review flag to approved, so @hallyn knows this PR is good to merge now?

solardiz commented:
@besser82 I'm not formally "approving" this for two reasons: (1) I am not a maintainer and (2) my review of changes to the legacy code wasn't sufficiently thorough and I don't feel like spending enough of my time on the legacy code that I don't use to perform its proper review and testing (and this really needs retesting of all settings, not just review).

ikerexxe reviewed (commented):
I have one last concern. Why have you included the changes on the `.po` files? Have you changed anything on those files?

besser82 commented:
> I have one last concern. Why have you included the changes on the `.po` files? Have you changed anything on those files?

No, but I updated the pot file, since my changes have added new translatable strings.  And as the pot file was updated a simple run of `make` now updates the po files as well.  Thus I've included them here, so other developers don't get annoyed by having them listed as modified files in their clone.

besser82 commented:
> > I have one last concern. Why have you included the changes on the `.po` files? Have you changed anything on those files?
> 
> No, but I updated the pot file, since my changes have added new translatable strings. And as the pot file was updated a simple run of `make` now updates the po files as well. Thus I've included them here, so other developers don't get annoyed by having them listed as modified files in their clone.

Anyways, I've dropped the update of the translation files from this PR, and will open a new PR for them.

ikerexxe commented:
> > > I have one last concern. Why have you included the changes on the `.po` files? Have you changed anything on those files?
> > 
> > 
> > No, but I updated the pot file, since my changes have added new translatable strings. And as the pot file was updated a simple run of `make` now updates the po files as well. Thus I've included them here, so other developers don't get annoyed by having them listed as modified files in their clone.
> 
> Anyways, I've dropped the update of the translation files from this PR, and will open a new PR for them.

Thanks. LGTM!

besser82 commented:
@hallyn This PR should be ready to merge, then.
